gui/button.3
.\"	$Csoft: button.3,v 1.27 2005/09/27 14:06:35 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 20, 2002
.Dt BUTTON 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm button
.Nd agar button widget
.Sh SYNOPSIS
.Bd -literal
#include <agar/gui/button.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget implements a simple push-button containing a string of text or
an image.
.Sh INHERITANCE HIERARCHY
.Pp
.Xr AG_ObjectAttach 3 ->
.Xr widget 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "AG_Button *"
.Fn AG_ButtonNew "void *parent" "char *caption"
.Pp
.Ft "AG_Button *"
.Fn AG_ButtonAct "void *parent" "char *caption" "u_int flags" "const char *fmt" "..."
.Pp
.Ft void
.Fn AG_ButtonInit "AG_Button *button" "char *caption" "u_int flags"
.Pp
.Ft void
.Fn AG_ButtonEnable "AG_Button *button"
.Pp
.Ft void
.Fn AG_ButtonDisable "AG_Button *button"
.Pp
.Ft void
.Fn AG_ButtonSetPadding "AG_Button *button" "int padding"
.Pp
.Ft void
.Fn AG_ButtonSetFocusable "AG_Button *button" "int focusable"
.Pp
.Ft void
.Fn AG_ButtonSetSticky "AG_Button *button" "int sticky"
.Pp
.Ft void
.Fn AG_ButtonSetJustification "AG_Button *button" "enum ag_button_justify jus"
.Pp
.Ft void
.Fn AG_ButtonSetSurface "AG_Button *button" "SDL_Surface *su"
.Pp
.Ft void
.Fn AG_ButtonSetRepeatMode "AG_Button *button" "int repeat_flag"
.Pp
.Ft void
.Fn AG_ButtonPrintf "AG_Button *button" "const char *fmt" "..."
.nr nS 0
.Pp
The
.Fn AG_ButtonNew
function allocates, initializes, and attaches a
.Nm
widget.
.Pp
The
.Fn AG_ButtonAct
variant creates a button and automatically binds the provided event handler
function to its
.Sq button-pushed
event.
The
.Fa fmt
and following arguments specify an
.Xr event 3
style argument list.
.Pp
The
.Fn AG_ButtonInit
function initializes a
.Nm
widget displaying the string
.Fa caption
(if not NULL).
Acceptable values for the
.Fa flags
argument include:
.Bd -literal
#define AG_BUTTON_INSENSITIVE	0x01	/* Not responsive */
#define AG_BUTTON_STICKY	0x02	/* Toggle state */
#define AG_BUTTON_REPEAT	0x08	/* Repeat button-pushed event */
#define AG_BUTTON_WFILL		0x10	/* Fill available width */
#define AG_BUTTON_HFILL		0x20	/* Fill available height */
#define AG_BUTTON_FOCUS		0x40	/* Focus button automatically */
.Ed
.Pp
Note that the
.Dv AG_BUTTON_FOCUS
flag is only effective when passed to
.Fn AG_ButtonAct ,
.Fn AG_ButtonInit
will ignore it.
.Pp
The
.Fn AG_ButtonEnable
and
.Fn AG_ButtonDisable
functions trigger the sensitivity of the
.Nm
widget.
.Pp
The
.Fn AG_ButtonSetPadding
function sets the padding around the label in pixels.
.Pp
The
.Fn AG_ButtonSetFocusable
function with an argument of 0 prevents the button from gaining focus.
The default is to allow buttons to gain focus.
.Pp
The
.Fn AG_ButtonSetSticky
function with an argument of 1 arranges for the button to trigger its state
only on reception of
.Fn window-mousebuttondown
events.
By default,
.Fn window-mousebuttonup
events reset the state to 0.
.Pp
The
.Fn AG_ButtonSetJustification
function controls the label justification setting:
.Pp
.Bd -literal
enum ag_button_justify {
	AG_BUTTON_LEFT,
	AG_BUTTON_CENTER,
	AG_BUTTON_RIGHT
};
.Ed
.Pp
The
.Fn AG_ButtonSetSurface
and
.Fn AG_ButtonPrintf
functions are used to modify the button label.
.Pp
The
.Fn AG_ButtonSetRepeatMode
flag enables or disables repeat mode.
Repeat mode causes multiple
.Sq button-pushed
events to be posted periodically for as long as the button is triggered.
Repeat mode is useful for things like
.Xr spinbutton 3 .
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -compact -width 25n
.It window-mousebuttonup
Release the button if the cursor is inside the widget area.
.It window-mousebuttondown
Press the button.
.It window-keyup
Release the button, unless sticky mode is in effect.
.It window-keydown
Press the button.
.It window-mousemotion
Abort a button press if the cursor is moving outside the button area and sticky
mode is not in effect.
.El
.Pp
The
.Nm
widget generates the following events:
.Pp
.Bl -tag -compact -width 2n
.It Fn button-pushed "int new_state"
The button was pressed.
The
.Fa new_state
argument is useful in conjunction with the
.Dv AG_BUTTON_STICKY
flag.
.It Fn button-mouseoverlap "int overlap"
The mouse cursor has entered/exited the button area.
This is typically is used to display tool tips.
.El
.Sh BINDINGS
The
.Nm
widget provides the following bindings:
.Pp
.Bl -tag -compact -width "int *state"
.It Va int *state
Button state (1=pressed, 0=released)
.El
.Sh EXAMPLES
The following code fragment creates a button and sets a handler function
for the
.Sq button-pushed
event:
.Bd -literal -offset indent
void
hello(AG_Event *event)
{
	char *s = AG_STRING(1);

	AG_TextMsg(AG_MSG_INFO, "Hello, %s!", s);
}
 
.Li ...

AG_Button *btn;

btn = AG_ButtonNew(parent, "Hello");
AG_SetEvent(btn, "button-pushed", hello, "%s", "world");
.Ed
.Sh SEE ALSO
.Xr agar 3 ,
.Xr widget 3 ,
.Xr window 3
.Sh HISTORY
The
.Nm
widget first appeared in Agar 1.0.
.Sh BUGS
A more general interface may be preferable to
.Fn AG_ButtonEnable
and
.Fn AG_ButtonDisable .
.Pp
This widget uses the
.Dv AG_WIDGET_UNFOCUSED_MOTION
flag which is very inefficient.


gui/button.c
/*	$Csoft: button.c,v 1.95 2005/10/01 14:15:38 vedge Exp $	*/

/*
 * Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <core/core.h>
#include <core/view.h>

#include <stdarg.h>

#include "button.h"

#include <gui/window.h>
#include <gui/primitive.h>
#include <gui/label.h>

const AG_WidgetOps agButtonOps = {
	{
		NULL,		/* init */
		NULL,		/* reinit */
		NULL,		/* destroy XXX */
		NULL,		/* load */
		NULL,		/* save */
		NULL		/* edit */
	},
	AG_ButtonDraw,
	AG_ButtonScale
};

static void button_mousemotion(AG_Event *);
static void button_mousebuttonup(AG_Event *);
static void button_mousebuttondown(AG_Event *);
static void button_keyup(AG_Event *);
static void button_keydown(AG_Event *);

AG_Button *
AG_ButtonNew(void *parent, const char *caption)
{
	AG_Button *btn;

	btn = Malloc(sizeof(AG_Button), M_OBJECT);
	AG_ButtonInit(btn, caption, 0);
	AG_ObjectAttach(parent, btn);
	return (btn);
}

AG_Button *
AG_ButtonAct(void *parent, const char *caption, u_int flags,
    void (*fn)(AG_Event *), const char *fmt, ...)
{
	AG_Button *btn;
	AG_Event *ev;

	btn = Malloc(sizeof(AG_Button), M_OBJECT);
	AG_ButtonInit(btn, caption, flags);
	AG_ObjectAttach(parent, btn);

	ev = AG_SetEvent(btn, "button-pushed", fn, NULL);
	AG_EVENT_GET_ARGS(ev, fmt);

	if (flags & AG_BUTTON_FOCUS) {
		AG_WidgetFocus(btn);
	}
	return (btn);
}


static Uint32
repeat_expire(void *obj, Uint32 ival, void *arg)
{
	AG_PostEvent(NULL, obj, "button-pushed", "%i", 1);
	return (agMouseSpinIval);
}

static Uint32
delay_expire(void *obj, Uint32 ival, void *arg)
{
	AG_Button *bu = obj;

	AG_ReplaceTimeout(bu, &bu->repeat_to, agMouseSpinIval);
	return (0);
}

void
AG_ButtonInit(AG_Button *bu, const char *caption, u_int flags)
{
	SDL_Surface *label;

	/* XXX replace the unfocused motion flag with a timer */
	AG_WidgetInit(bu, "button", &agButtonOps,
	    AG_WIDGET_FOCUSABLE|AG_WIDGET_UNFOCUSED_MOTION);
	AG_WidgetBind(bu, "state", AG_WIDGET_BOOL, &bu->state);

	label = (caption == NULL) ? NULL :
	    AG_TextRender(NULL, -1, AG_COLOR(BUTTON_TXT_COLOR), caption);
	AG_WidgetMapSurface(bu, label);

	bu->flags = 0;
	bu->state = 0;
	bu->justify = AG_BUTTON_CENTER;
	bu->padding = 4;

	AG_SetTimeout(&bu->repeat_to, repeat_expire, NULL, 0);
	AG_SetTimeout(&bu->delay_to, delay_expire, NULL, 0);

	AG_SetEvent(bu, "window-mousebuttonup", button_mousebuttonup, NULL);
	AG_SetEvent(bu, "window-mousebuttondown", button_mousebuttondown, NULL);
	AG_SetEvent(bu, "window-mousemotion", button_mousemotion, NULL);
	AG_SetEvent(bu, "window-keyup", button_keyup, NULL);
	AG_SetEvent(bu, "window-keydown", button_keydown, NULL);

	if (flags & AG_BUTTON_WFILL) { AGWIDGET(bu)->flags |= AG_WIDGET_WFILL; }
	if (flags & AG_BUTTON_HFILL) { AGWIDGET(bu)->flags |= AG_WIDGET_HFILL; }
}

void
AG_ButtonScale(void *p, int w, int h)
{
	AG_Button *bu = p;
	SDL_Surface *label = AGWIDGET_SURFACE(bu,0);

	if (w == -1 && h == -1) {
		if (label != NULL) {
			AGWIDGET(bu)->w = label->w + bu->padding*2;
			AGWIDGET(bu)->h = label->h + bu->padding*2;
		} else {
			AGWIDGET(bu)->w = 1;
			AGWIDGET(bu)->h = 1;
		}
	}
}

void
AG_ButtonDraw(void *p)
{
	AG_Button *bu = p;
	SDL_Surface *label = AGWIDGET_SURFACE(bu,0);
	int x = 0, y = 0;
	int pressed;
	
	if (AGWIDGET(bu)->w < 8 || AGWIDGET(bu)->h < 8)
		return;

	pressed = AG_WidgetBool(bu, "state");
	if (bu->flags & AG_BUTTON_INSENSITIVE) {
		agPrim.box(bu,
		    0, 0,
		    AGWIDGET(bu)->w, AGWIDGET(bu)->h,
		    -1,
		    AG_COLOR(BUTTON_DIS_COLOR));
	} else {
		agPrim.box(bu,
		    0, 0,
		    AGWIDGET(bu)->w, AGWIDGET(bu)->h,
		    pressed ? -1 : 1,
		    AG_COLOR(BUTTON_COLOR));
	}

	if (label != NULL) {
		switch (bu->justify) {
		case AG_BUTTON_LEFT:
			x = bu->padding;
			break;
		case AG_BUTTON_CENTER:
			x = AGWIDGET(bu)->w/2 - label->w/2;
			break;
		case AG_BUTTON_RIGHT:
			x = AGWIDGET(bu)->w - label->w - bu->padding;
			break;
		}
		y = ((AGWIDGET(bu)->h - label->h)/2) - 1;	/* Middle */

		if (pressed) {
			x++;
			y++;
		}
		AG_WidgetBlitSurface(bu, 0, x, y);
	}
}

static void
button_mousemotion(AG_Event *event)
{
	AG_Button *bu = AG_SELF();
	AG_WidgetBinding *stateb;
	int x = AG_INT(1);
	int y = AG_INT(2);
	int *pressed;

	if (bu->flags & AG_BUTTON_INSENSITIVE)
		return;

	stateb = AG_WidgetGetBinding(bu, "state", &pressed);
	if (!AG_WidgetRelativeArea(bu, x, y)) {
		if ((bu->flags & AG_BUTTON_STICKY) == 0
		    && *pressed == 1) {
			*pressed = 0;
			AG_WidgetBindingChanged(stateb);
		}
		if (bu->flags & AG_BUTTON_MOUSEOVER) {
			bu->flags &= ~(AG_BUTTON_MOUSEOVER);
			AG_PostEvent(NULL, bu, "button-mouseoverlap", "%i", 0);
		}
	} else {
		bu->flags |= AG_BUTTON_MOUSEOVER;
		AG_PostEvent(NULL, bu, "button-mouseoverlap", "%i", 1);
	}
	AG_WidgetUnlockBinding(stateb);
}

static void
button_mousebuttondown(AG_Event *event)
{
	AG_Button *bu = AG_SELF();
	int button = AG_INT(1);
	AG_WidgetBinding *stateb;
	int *pushed;
	
	if (bu->flags & AG_BUTTON_INSENSITIVE)
		return;

	AG_WidgetFocus(bu);

	if (button != SDL_BUTTON_LEFT)
		return;
	
	stateb = AG_WidgetGetBinding(bu, "state", &pushed);
	if (!(bu->flags & AG_BUTTON_STICKY)) {
		*pushed = 1;
	} else {
		*pushed = !(*pushed);
		AG_PostEvent(NULL, bu, "button-pushed", "%i", *pushed);
	}
	AG_WidgetBindingChanged(stateb);
	AG_WidgetUnlockBinding(stateb);

	if (bu->flags & AG_BUTTON_REPEAT) {
		AG_DelTimeout(bu, &bu->repeat_to);
		AG_ReplaceTimeout(bu, &bu->delay_to, agMouseSpinDelay);
	}
}

static void
button_mousebuttonup(AG_Event *event)
{
	AG_Button *bu = AG_SELF();
	int button = AG_INT(1);
	AG_WidgetBinding *stateb;
	int *pushed;
	int x = AG_INT(2);
	int y = AG_INT(3);
		
	if (bu->flags & AG_BUTTON_REPEAT) {
		AG_DelTimeout(bu, &bu->repeat_to);
		AG_DelTimeout(bu, &bu->delay_to);
	}
	
	if ((bu->flags & AG_BUTTON_INSENSITIVE) ||
	    x < 0 || y < 0 ||
	    x > AGWIDGET(bu)->w || y > AGWIDGET(bu)->h) {
		return;
	}
	
	stateb = AG_WidgetGetBinding(bu, "state", &pushed);
	if (*pushed &&
	    button == SDL_BUTTON_LEFT &&
	    !(bu->flags & AG_BUTTON_STICKY)) {
	    	*pushed = 0;
		AG_PostEvent(NULL, bu, "button-pushed", "%i", *pushed);
		AG_WidgetBindingChanged(stateb);
	}
	AG_WidgetUnlockBinding(stateb);
}

static void
button_keydown(AG_Event *event)
{
	AG_Button *bu = AG_SELF();
	int keysym = AG_INT(1);
	
	if (bu->flags & AG_BUTTON_INSENSITIVE)
		return;

	if (keysym == SDLK_RETURN || keysym == SDLK_SPACE) {
		AG_WidgetSetBool(bu, "state", 1);
		AG_PostEvent(NULL, bu, "button-pushed", "%i", 1);

		if (bu->flags & AG_BUTTON_REPEAT) {
			AG_DelTimeout(bu, &bu->repeat_to);
			AG_ReplaceTimeout(bu, &bu->delay_to, 800);
		}
	}
}

static void
button_keyup(AG_Event *event)
{
	AG_Button *bu = AG_SELF();
	int keysym = AG_INT(1);
	
	if (bu->flags & AG_BUTTON_INSENSITIVE)
		return;
	
	if (bu->flags & AG_BUTTON_REPEAT) {
		AG_DelTimeout(bu, &bu->delay_to);
		AG_DelTimeout(bu, &bu->repeat_to);
	}

	if (keysym == SDLK_RETURN || keysym == SDLK_SPACE) {
		AG_WidgetSetBool(bu, "state", 0);
		AG_PostEvent(NULL, bu, "button-pushed", "%i", 0);
	}
}

void
AG_ButtonEnable(AG_Button *bu)
{
	bu->flags &= ~(AG_BUTTON_INSENSITIVE);
}

void
AG_ButtonDisable(AG_Button *bu)
{
	bu->flags |= (AG_BUTTON_INSENSITIVE);
}

void
AG_ButtonSetPadding(AG_Button *bu, int padding)
{
	bu->padding = padding;
}

void
AG_ButtonSetFocusable(AG_Button *bu, int focusable)
{
	if (focusable) {
		AGWIDGET(bu)->flags |= AG_WIDGET_FOCUSABLE;
		AGWIDGET(bu)->flags &= ~(AG_WIDGET_UNFOCUSED_BUTTONUP);
	} else {
		AGWIDGET(bu)->flags &= ~(AG_WIDGET_FOCUSABLE);
		AGWIDGET(bu)->flags |= AG_WIDGET_UNFOCUSED_BUTTONUP;
	}
}

void
AG_ButtonSetSticky(AG_Button *bu, int sticky)
{
	if (sticky) {
		bu->flags |= (AG_BUTTON_STICKY);
	} else {
		bu->flags &= ~(AG_BUTTON_STICKY);
	}
}

void
AG_ButtonSetJustification(AG_Button *bu, enum ag_button_justify jus)
{
	bu->justify = jus;
}

void
AG_ButtonSetSurface(AG_Button *bu, SDL_Surface *su)
{
	AG_WidgetReplaceSurface(bu, 0, su);
}

void
AG_ButtonSetRepeatMode(AG_Button *bu, int repeat)
{
	if (repeat) {
		bu->flags |= (AG_BUTTON_REPEAT);
	} else {
		AG_DelTimeout(bu, &bu->repeat_to);
		AG_DelTimeout(bu, &bu->delay_to);
		bu->flags &= ~(AG_BUTTON_REPEAT);
	}
}

void
AG_ButtonPrintf(AG_Button *bu, const char *fmt, ...)
{
	char buf[AG_LABEL_MAX];
	va_list args;

	va_start(args, fmt);
	vsnprintf(buf, sizeof(buf), fmt, args);
	va_end(args);

	AG_WidgetReplaceSurface(bu, 0,
	    AG_TextRender(NULL, -1, AG_COLOR(BUTTON_TXT_COLOR), buf));
}

